src/sets.jl

```diff
@@ -240,3 +239,229 @@ function distance_to_set(d::DefaultDistance, v::AbstractVector{T}, s::MOI.Indica
         (distance_to_set(d, z, MOI.ZeroOne()), distance_to_set(v[2], s.set))
     )
 end
+
+
+# find expression of projections on cones and their derivatives here:
+#   https://stanford.edu/~boyd/papers/pdf/cone_prog_refine.pdf
+
+
+"""
+    projection_on_set(::AbstractDistance, ::MOI.Zeros, z::Array{Float64}, dual=true)
+
+projection of vector `z` on zero cone i.e. K = {0} or its dual
+"""
+function projection_on_set(::DefaultDistance, ::MOI.Zeros, z::Array{Float64}, dual=true)
+    return dual ? z : zeros(Float64, size(z))
+end
+
+function projection_on_set(::DefaultDistance, ::MOI.EqualTo, z::Array{Float64}, dual=true)
+    return dual ? z : zeros(Float64, size(z))
+end
+
+function projection_on_set(::DefaultDistance, ::MOI.EqualTo, z::Float64, dual=true)
+    return dual ? [z] : zeros(Float64, 1)
+end
+
+"""
+    projection of vector `z` on Nonnegative cone i.e. K = R+
+"""
+function projection_on_set(::DefaultDistance, ::MOI.Nonnegatives, z::Array{Float64})
+    return max.(z, 0.0)
+end
+
+"""
+    projection of vector `z` on second order cone i.e. K = {(t, x) ∈ R+ × Rn |  ||x|| ≤ t }
+"""
+function projection_on_set(::DefaultDistance, ::MOI.SecondOrderCone, z::Array{Float64})
+    t = z[1]
+    x = z[2:length(z)]
+    norm_x = norm(x)
+    if norm_x <= t
+        return copy(z)
+    elseif norm_x <= -t
+        return zeros(Float64, size(z))
+    else
+        result = zeros(Float64, size(z))
+        result[1] = 1.0
+        result[2:length(z)] = x / norm_x
+        result *= (norm_x + t) / 2.0
+        return result
+    end
+end
+
+"""
+    projection of vector `z` on positive semidefinite cone i.e. K = S^n⨥
+"""
+function projection_on_set(::DefaultDistance, ::MOI.PositiveSemidefiniteConeTriangle, z::Array{Float64})
+    dim = Int64(floor(√(2*length(z))))
+    X = unvec_symm(z, dim)
+    λ, U = eigen(X)
+    return vec_symm(U * Diagonal(max.(λ,0)) * U')
+end
+
+"""
+    Returns a dim-by-dim symmetric matrix corresponding to `x`.
+
+    `x` is a vector of length dim*(dim + 1)/2, corresponding to a symmetric
+    matrix; the correspondence is as in SCS.
+    X = [ X11 X12 ... X1k
+          X21 X22 ... X2k
+          ...
+          Xk1 Xk2 ... Xkk ],
+    where
+    vec(X) = (X11, sqrt(2)*X21, ..., sqrt(2)*Xk1, X22, sqrt(2)*X32, ..., Xkk)
+"""
+function unvec_symm(x, dim)
+    X = zeros(dim, dim)
+    for i in 1:dim
+        for j in i:dim
+            X[j,i] = x[(i-1)*dim-Int(((i-1)*i)/2)+j]
+        end
+    end
+    X = X + X'
```

Comment:
`X = X + X'` this could be replaced by setting at the same time X[i,j] and X[j,i] in the loop above

Reply:
also, whenever you know the indices will be respected `@inbounds` in front of the for loop or indexing operation
https://docs.julialang.org/en/v1/devdocs/boundscheck/

Reply:
ok. i'll keep it in mind